Rewrite User storage and UUID cache

JRoy opened the pull request:
A rewrite of user loading, user caching, uuid caching, and user memory management.

Closes #4512
Closes #4342
Closes #4566
Closes #4581

---

### Technical Breakdown of `ModernUserMap` + `ModernUUIDCache`

TODO :trollface: 

---

Checklist;
- [x] Store username history more sanely
- [x] Rewrite the UUID cache
- [x] Rewrite the UserMap and replace the old UserMap
- [x] Add ABI backwards compatibility for certain hot methods in old UserMap

JRoy committed:
Store username history inside userdata

JRoy committed:
Remove legacy uuidconvert and uuidtest subcommands

JRoy committed:
Remove legacy code from UserMap and UUIDMap

JRoy committed:
Start work on ModernUUIDCache

JRoy committed:
Also cache all uuids in seperate file

JRoy committed:
Complete first iteration of ModernUserMap

JRoy committed:
Replace uuid cache upgrade

JRoy committed:
Replace all usages of UserMap with ModernUserMap

JRoy committed:
Backport some select methods to the old UserMap class

JRoy committed:
Forcefully save NPC names into uuid cache

JRoy committed:
Properly save the UUID/Name cache on shutdown

JRoy committed:
Actually read more than one user/uuid

JRoy committed:
Add /ess usermap command for information on the usermap

JRoy committed:
Add usermap.bin and uuids.bin to .gitignore

JRoy committed:
Make StringUtil#safeString null-safe

JRoy committed:
Move config save hook to contructor

JRoy committed:
Initialize legacy user map

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/usermap-rewrite

tadhgboyle commented:
:shipit:

pop4959 reviewed (dismissed):
//noinspection ConstantConditions

JRoy commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUUIDCache.java:
```diff
@@ -0,0 +1,236 @@
+     * * mdcfe doesn't log in the server for 31 days
+     * * JRoy changes his name to mdcfeYT420
+     * * mdcfeYT420 (previously JRoy) logs in the server
+     * * In a UUID->Name based map, multiple uuids now point to the same map
```
```suggestion
     * In a UUID->Name based map, multiple uuids now point to the same map
```

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUUIDCache.java

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandseen.java

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

mdcfe commented:
Bumped to 2.20 as per internal discussions

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/IUser.java
#	Essentials/src/main/java/com/earth2me/essentials/User.java

Gilthunderx commented:
PULLL!

darbyjack commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUUIDCache.java:
```diff
@@ -0,0 +1,236 @@
+     * preventing any command which allows for offline players to resolve a
+     * single uuid from a name.
+     *
+     * This map is baked by a file-based cache. If this cache is missing, a new
```
backed*?

JRoy replied:
```suggestion
     * This map is backed by a file-based cache. If this cache is missing, a new
```

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Update Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUUIDCache.java

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

triagonal committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/UserData.java
#	Essentials/src/main/java/com/earth2me/essentials/config/holders/UserConfigHolder.java
#	providers/1_12Provider/src/main/java/com/earth2me/essentials/OfflinePlayer.java

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Fix compile error

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/UserData.java

JRoy committed:
Add a few more debug commands

JRoy committed:
Fix usermap not getting saved during new uuid insertions

JRoy committed:
Add some additional debug messages

JRoy committed:
Prefer UUIDv4 during usermap initialization

JRoy committed:
Add more usermap debug commands

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/EssentialsUpgrade.java
#	Essentials/src/main/java/com/earth2me/essentials/UUIDMap.java
#	Essentials/src/main/java/com/earth2me/essentials/config/EssentialsConfiguration.java
#	Essentials/src/main/java/com/earth2me/essentials/economy/vault/VaultEconomyProvider.java

JRoy committed:
joker moment

JRoy committed:
...

JRoy committed:
Fix a few typos in javadocs

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Compare logout timestamps when replacing same uuid version

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Don't cache users into map for luckperms contexts

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

# Conflicts:
#	Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java

JRoy committed:
Merge remote-tracking branch 'upstream/2.x' into refactor/usermap-rewrite

JRoy committed:
Always update usernames on join

Always call ModernUserMap#getUser(Player) during join to ensure the
fetched/created User object has the most up-to-date name on the underlying
Player object. Additionally, ensure that ModernUUIDCache#updateCache is always
called during #getUser and not just when we generate a new User object (which
always never happens by the time the call is made :skull:).

JRoy committed:
Add ModernUserMap#loadUncachedUser(Player)

Used to create User objects based on player objects for use from when before
Bukkit places the user into its own internal usermap.
Without this, ModernUserMap#loadUncachedUser(UUID) would attempt to fetch
the player from the server only to not return null because we first try to
create a user during PlayerLoginEvent which is before Bukkit puts the player
into its internal usermap.

JRoy committed:
if i broke one more damn thing, i'm reverting everything

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:
```diff
@@ -527,6 +528,7 @@ public void onPlayerLoginBanned(final PlayerLoginEvent event) {
     public void onPlayerLogin(final PlayerLoginEvent event) {
         if (event.getResult() == Result.KICK_FULL) {
             final User kfuser = ess.getUser(event.getPlayer());
+            kfuser.update(event.getPlayer());
```
hmmmm

JRoy replied:
i'm not touching these calls, it broke when i removed them and worked when i added them

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/api/Economy.java:
```diff
@@ -94,7 +96,7 @@ private static User getUserByName(final String name) {
 
         if (user == null) {
-            user = getUserByUUID(UUID.nameUUIDFromBytes(("NPC:" + StringUtil.safeString(name)).getBytes(Charsets.UTF_8)));
+            user = getUserByUUID(UUID.nameUUIDFromBytes(("NPC:" + (ess.getSettings().isSafeUsermap() ? StringUtil.safeString(name) : name)).getBytes(Charsets.UTF_8)));
```
This is likely to break plugin accounts (eg Towny) on servers that previously disabled username sanitisation?

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

JRoy committed:
Merge branch '2.x' into refactor/usermap-rewrite

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/economy/vault/VaultEconomyProvider.java:
```diff
@@ -79,7 +80,7 @@ public boolean hasAccount(String playerName) {
         }
         // We may not have the player name in the usermap, let's double check an NPC account with this name doesn't exist.
-        return com.earth2me.essentials.api.Economy.playerExists(UUID.nameUUIDFromBytes(("NPC:" + StringUtil.safeString(playerName)).getBytes(Charsets.UTF_8)));
+        return com.earth2me.essentials.api.Economy.playerExists(UUID.nameUUIDFromBytes(("NPC:" + (ess.getSettings().isSafeUsermap() ? StringUtil.safeString(playerName) : playerName)).getBytes(Charsets.UTF_8)));
```
Same as before, potential to break migrating servers with `use-safe-usermap: false`?

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUUIDCache.java:
```diff
@@ -0,0 +1,250 @@
+        saveNameToUuidCache();
+    }
+
+    public static void writeUuidCache(final File file, Set<UUID> uuids) throws IOException {
+        try (final DataOutputStream dos = new DataOutputStream(new FileOutputStream(file))) {
+            for (final UUID uuid: uuids) {
+                dos.writeLong(uuid.getMostSignificantBits());
+                dos.writeLong(uuid.getLeastSignificantBits());
+            }
+        }
+    }
+
+    public static void writeNameUuidMap(final File file, final Map<String, UUID> nameToUuidMap) throws IOException {
+        try (final DataOutputStream dos = new DataOutputStream(new FileOutputStream(file))) {
+            for (final Map.Entry<String, UUID> entry : nameToUuidMap.entrySet()) {
+                dos.writeUTF(entry.getKey());
+                final UUID uuid = entry.getValue();
+                dos.writeLong(uuid.getMostSignificantBits());
+                dos.writeLong(uuid.getLeastSignificantBits());
+            }
+        }
+    }
```
Do these need to be public? ~~(and how much of this class needs to be public?)~~

JRoy replied:
They're used in EssentialsUpgrade

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUserMap.java:
```diff
@@ -0,0 +1,170 @@
+            ess.getLogger().log(Level.INFO, "Essentials updated the underlying Player object for " + user.getUUID());
+            user.update(base);
+        }
+        uuidCache.updateCache(user.getUUID(), user.getName());
```
Do we definitely always want to update the UUID cache here?

JRoy replied:
yeah thanks for asking

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUserMap.java:
```diff
@@ -0,0 +1,170 @@
+        Player player = ess.getServer().getPlayer(uuid);
+        if (player != null) {
+            user = new User(player, ess);
+            uuidCache.updateCache(uuid, player.getName());
```
```suggestion
            // This is a real player, cache their UUID
            uuidCache.updateCache(uuid, player.getName());
```

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUserMap.java:
```diff
@@ -0,0 +1,170 @@
+        this.userCache = CacheBuilder.newBuilder()
+                .maximumSize(ess.getSettings().getMaxUserCacheCount())
+                .softValues()
+                .build(this);
```
Not sure about this - do we actually want `load` exposed publicly?

JRoy replied:
not a big deal buddy

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/userstorage/ModernUserMap.java:
```diff
@@ -0,0 +1,170 @@
+            ess.getLogger().log(Level.INFO, "Essentials created a User for " + base.getName() + " (" + base.getUniqueId() + ") for non Bukkit type: " + base.getClass().getName());
+            user = new User(base, ess);
+        } else if (!base.equals(user.getBase())) {
+            ess.getLogger().log(Level.INFO, "Essentials updated the underlying Player object for " + user.getUUID());
```
In theory would break online users if some other plugin gives us their own `Player` type for an online player for some reason. Not sure if there's any cases where we would want to guard against this?

JRoy replied:
why would they do that

mdcfe commented on Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:
```diff
@@ -577,52 +575,77 @@ private void runHomes(final Server server, final CommandSource sender, final Str
+                final Set<UUID> uuids = new HashSet<>(ess.getUsers().getAllUserUUIDs());
+                ess.runTaskAsynchronously(() -> {
+                    final File userdataFolder = new File(ess.getDataFolder(), "userdata");
+                    final File backupFolder = new File(ess.getDataFolder(), "userdata-npc-backup-boogaloo");
```
Include the date/time in the folder name? Would make it easier for admins to identify if they did it multiple times

mdcfe reviewed (changes requested):
Side note: do we want an API interface for `ModernUserMap`? Just so there's an explicit (and limited) set of methods that plugins should expect to use, otherwise I anticipate people will just call internal methods

JRoy committed:
Use safe string for Economy and VaultEconomyProvider always

JRoy committed:
Change some method visibility, create IUserMap

JRoy committed:
Misc review items

mdcfe commented on .idea/checkstyle-idea.xml:
```diff
@@ -7,13 +7,13 @@
         <ConfigurationLocation id="bundled-sun-checks" type="BUNDLED" scope="All" description="Sun Checks">(bundled)</ConfigurationLocation>
         <ConfigurationLocation id="bundled-google-checks" type="BUNDLED" scope="All" description="Google Checks">(bundled)</ConfigurationLocation>
-        <ConfigurationLocation id="321e5753-4201-4d2c-bf31-6232011122fd" type="PROJECT_RELATIVE" scope="All" description="EssentialsX">
+        <ConfigurationLocation id="9c559e66-f9d0-4553-b942-bd6095fd8a13" type="PROJECT_RELATIVE" scope="All" description="EssentialsX">
```
Don't need to commit this

mdcfe committed:
Merge branch 'dev/2.20' into refactor/usermap-rewrite

JRoy committed:
Restore checkstyle-idea.xml

JRoy committed:
Add npc-name property for generated NPC accounts

mibby commented:
@JRoy Not entirely sure where to leave feedback for this but noticed that every time Citizens loads an npc (server restarts, teleporting around), Essentials spams creating a user for them.
https://paste.gg/p/anonymous/e7fd399e9db6436d8ed3c5ee15dae70a/files/febeaa22bd6846a8aa7d8a93acb63b85/raw

EssentialsX-2.20.0-dev+5
https://ci.ender.zone/view/All/job/EssentialsX-Experimental/100007/
Citizens2 dev 2705
https://ci.citizensnpcs.co/job/Citizens2/

JeffP07 commented:
> @JRoy Not entirely sure where to leave feedback for this but noticed that every time Citizens loads an npc (server restarts, teleporting around), Essentials spams creating a user for them. https://paste.gg/p/anonymous/e7fd399e9db6436d8ed3c5ee15dae70a/files/febeaa22bd6846a8aa7d8a93acb63b85/raw
> 
> EssentialsX-2.20.0-dev+5 https://ci.ender.zone/view/All/job/EssentialsX-Experimental/100007/ Citizens2 dev 2705 https://ci.citizensnpcs.co/job/Citizens2/

I had this issue and found it was caused by jail.yml having old data that was formatted in a different way than new jails. Since I don't actually use the jails, removing the file contents worked for me.
